pam/integration-tests/gdm-module-handler_test.go

```diff
@@ -0,0 +1,239 @@
+		})
+		if idx < 0 {
+			return fmt.Errorf("unknown broker: %s", ev.BrokerSelected.BrokerId)
+		}
+		gh.brokerID = gh.brokersInfos[idx].Id
+		gh.t.Logf("Using broker '%s'", gh.brokersInfos[idx].Name)
+		require.Equal(gh.t, gh.selectedBrokerName, gh.brokersInfos[idx].Name,
+			"Selected broker name does not match expected one")
+
+	case *gdm.EventData_AuthModesReceived:
+		gh.authModes = ev.AuthModesReceived.AuthModes
+
+	case *gdm.EventData_AuthModeSelected:
+		gh.authModeID = ev.AuthModeSelected.AuthModeId
+
+	case *gdm.EventData_UiLayoutReceived:
+		layout := ev.UiLayoutReceived.UiLayout
+		if layout.Label != nil {
+			gh.t.Logf("%s:", *layout.Label)
+		}
+
+	case *gdm.EventData_StartAuthentication:
+		idx := slices.IndexFunc(gh.authModes, func(mode *authd.GAMResponse_AuthenticationMode) bool {
+			return mode.Id == gh.authModeID
+		})
+		if idx < 0 {
+			return fmt.Errorf("unknown auth mode type: %s", gh.authModeID)
+		}
+		if len(gh.selectedAuthModeIDs) < 1 {
+			return fmt.Errorf("unexpected authentication started with mode '%s', we've nothing to reply",
+				gh.authModeID)
+		}
+		require.Equal(gh.t, gh.selectedAuthModeIDs[0], gh.authModes[idx].Id,
+			"Selected authentication mode ID does not match expected one")
+		gh.selectedAuthModeIDs = slices.Delete(gh.selectedAuthModeIDs, 0, 1)
+
+	case *gdm.EventData_AuthEvent:
+		gh.t.Logf("Authentication event: %s", ev.AuthEvent.Response)
+		if msg := ev.AuthEvent.Response.Msg; msg != "" {
+			gh.t.Logf("Got message: %s", msg)
+		}
+	}
+	return nil
+}
+
+func (gh *gdmTestModuleHandler) exampleHandleAuthDRequest(gdmData *gdm.Data) (*gdm.Data, error) {
+	switch req := gdmData.Request.Data.(type) {
+	case *gdm.RequestData_UiLayoutCapabilities:
+		return &gdm.Data{
+			Type: gdm.DataType_response,
+			Response: &gdm.ResponseData{
+				Type: gdmData.Request.Type,
+				Data: &gdm.ResponseData_UiLayoutCapabilities{
+					UiLayoutCapabilities: &gdm.Responses_UiLayoutCapabilities{
+						SupportedUiLayouts: gh.supportedLayouts,
+					},
+				},
+			},
+		}, nil
+
+	case *gdm.RequestData_ChangeStage:
+		if gdmData.Request.Data == nil {
+			return nil, fmt.Errorf("missing stage data")
+		}
+		gh.currentStage = req.ChangeStage.Stage
+		log.Debugf(context.TODO(), "Switching to stage %d", gh.currentStage)
+
+		return &gdm.Data{
+			Type: gdm.DataType_response,
+			Response: &gdm.ResponseData{
+				Type: gdmData.Request.Type,
+				Data: &gdm.ResponseData_Ack{},
+			},
+		}, nil
+
+	default:
+		return nil, fmt.Errorf("unknown request type")
+	}
+}
+
+// RespondPAMBinary is a dummy conversation callback adapter to implement [pam.BinaryPointerConversationFunc].
+func (gh *gdmTestModuleHandler) RespondPAMBinary(ptr pam.BinaryPointer) (pam.BinaryPointer, error) {
+	return gdm.DataConversationFunc(func(inData *gdm.Data) (*gdm.Data, error) {
+		outData, err := gh.exampleHandleGdmData(inData)
+		if err != nil {
+			json, jsonErr := inData.JSON()
+			require.NoError(gh.t, jsonErr)
```

Comment:
There is still this one (and more below), missing the intent of the change, also, are we in setup/teardown/the test itself?

Reply:
Oh... that was missed... I hope I've added descriptions to them all now, if grep is not wrong.